### PR TITLE
Harvest Strategy

### PR DESCRIPTION
**Pull request by idlehusbandry**

### CODE REVIEW

File: contracts/strategies/harvest/IdleHarvestStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+            underlyingToken.safeTransferFrom(msg.sender, address(this), _amount);
+            lastIndexAmount = lastIndexAmount + _amount;
+            minted = _depositToVault(_amount);
+        }
+    }
+
+    /// @notice internal function to deposit the funds to the vault
+    /// @param _amount Amount of tokens to deposit
+    function _depositToVault(uint256 _amount) internal returns (uint256) {
+        underlyingToken.safeApprove(strategyToken, _amount);
```
Comment (massun-onibakuchi) [inline contracts/strategies/harvest/IdleHarvestStrategy.sol:235]: should avoid reading a storage variable repeately.
`address _strategyToken = strategyToken;`

**Review by massun-onibakuchi — CHANGES_REQUESTED** [review]

File: contracts/strategies/harvest/IdleHarvestStrategy.sol (hunk 1 above)
Comment (bugduino) [reply contracts/strategies/harvest/IdleHarvestStrategy.sol:212]: +1

File: hardhat.config.js [hunk 2]
```diff
@@ -27,62 +27,59 @@ module.exports = {
...
+        // blockNumber: 13086034 // no stkAAVE in the contract (for test-harvest task)
+        // blockNumber: 13126332 // there are stkAAVE in the contract in cooldown
+        // blockNumber: 13261760 // pre transfer ownership
+        // blockNumber: 13666020 // convex/lido integration tests
+        // blockNumber: 13728440 // idleFEI upgraded
+        // blockNumber: 13739407 // idleDAI upgraded
+        // blockNumber: 13776718 // lido deploy
+        // blockNumber: 13810230 // convex deploy
+        // blockNumber: 14075568 // StakingRewards deploy
+        blockNumber: 14141000, // harvest testing
```
Comment (bugduino) [inline hardhat.config.js:81]: Please keep only this block as change for this file and revert the rest

File: package.json [hunk 3]
```diff
@@ -25,7 +25,7 @@
   },
   "homepage": "https://idle.finance",
   "scripts": {
-    "test": "npx hardhat test test/*.js",
+    "test": "npx hardhat test --network hardhat",
```
Comment (bugduino) [inline package.json:28]: `hardhat` is the default network and `test/*.js` it's on purpose so if you need to add in an integration test in fork launch it with `yarn test-integration`

File: test/IdleMStableStrategy.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
+require("hardhat/config");
```
Comment (bugduino) [inline test/IdleMStableStrategy.js:1]: Why is this file in this PR?

File: test/integration/mstable/IdleMStableStrategy.js [hunk 5]
```diff
@@ -49,7 +49,7 @@ const AMOUNT_TO_TRANSFER = BN("10000000000000000000000"); // 10k
 // const KEY_SAVINGS_MANAGER = "0x12fe936c77a1e196473c4314f3bed8eeac1d757b319abb85bdda70df35511bf1";
 const savingsManagerAddress = "0xBC3B550E0349D74bF5148D86114A48C3B4Aa856F";
 
-describe.only("IdleMStableStrategy", function () {
```
Comment (bugduino) [inline test/integration/mstable/IdleMStableStrategy.js:52]: Why this change here?

File: test/integration/mstable/IdleMStableStrategyWrapper.js [hunk 6]
```diff
@@ -24,9 +24,10 @@ const DAIAddress = "0x6B175474E89094C44Da98b954EedeAC495271d0F";
 
 const dai_whale = "0xE78388b4CE79068e89Bf8aA7f218eF6b9AB0e9d0";
```
Comment (bugduino) [inline test/integration/mstable/IdleMStableStrategyWrapper.js:25]: Please revert all changes not related to the Harvest strategy

**Review by bugduino — COMMENTED** [review]
Added some comments. In general the strategy seems ok. Tests need assertions because right now there are none basically

File: test/integration/mstable/IdleMStableStrategy.js (hunk 5 above)
Comment (idlehusbandry) [reply test/integration/mstable/IdleMStableStrategy.js:52]: I am running the mstable tests in harvest branch, hence remove `.only`

Comment (idlehusbandry) [reply test/integration/mstable/IdleMStableStrategy.js:52]: The tests only run when block number is set correctly. Hence I had to stop mstable specific tests for harvest contracts.

File: test/IdleMStableStrategy.js (hunk 4 above)
Comment (idlehusbandry) [reply test/IdleMStableStrategy.js:1]: The integration file got copied. It shouldn't have been there.
I have removed it.

File: package.json (hunk 3 above)
Comment (idlehusbandry) [reply package.json:28]: Using the same now.

File: test/IdleMStableStrategy.js [hunk 7]
```diff
@@ -1,193 +0,0 @@
-require("hardhat/config");
```
Comment (bugduino) [inline test/IdleMStableStrategy.js]: This file should not be deleted, just not touched in this branch

File: hardhat.config.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
-require('dotenv').config();
```
Comment (bugduino) [inline hardhat.config.js:1]: Again all changes should be reverted here besides the addition of the `blockNumber` used for tests

File: package.json [hunk 9]
```diff
@@ -25,7 +25,7 @@
   },
   "homepage": "https://idle.finance",
   "scripts": {
-    "test": "npx hardhat test test/*.js",
```
Comment (bugduino) [inline package.json:28]: This should be reverted

File: contracts/strategies/harvest/IdleHarvestStrategy.sol [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    /// @notice redeem the rewards. Claims reward as per the _extraData
+    /// @return rewards amount of reward that is deposited to vault
+    function redeemRewards(bytes calldata) external override onlyIdleCDO returns (uint256[] memory rewards) {
+        rewards = _redeemRewards();
+    }
+
+    /// @notice internal function to claim the rewards
+    function _redeemRewards() internal returns (uint256[] memory rewards) {
```
Comment (bugduino) [inline contracts/strategies/harvest/IdleHarvestStrategy.sol:111]: I would merge the `redeemRewards` and `_redeemRewards` into a single method given that it's used only in the external `redeemRewards` now

File: contracts/strategies/harvest/IdleHarvestStrategy.sol (hunk 1 above)
Comment (bugduino) [reply contracts/strategies/harvest/IdleHarvestStrategy.sol:205]: this is not resolved

File: test/integration/harvest/IdleHarvestStrategy_DAI.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+  });
+
+  afterEach(async () => {
+    await hre.network.provider.request({
+      method: "evm_revert",
+      params: [snapshotId],
+    });
+  });
+
+  it("Deposit", async () => {
```
Comment (bugduino) [inline test/integration/harvest/IdleHarvestStrategy_DAI.js:101]: Each test should have at least an assertion

**Review by bugduino — CHANGES_REQUESTED** [review]
There are several things that was already mentioned that are not yet resolved. 
- Tests are missing assertions. Each test should have at least one assertion
- changes in hardhat.config.js should be reverted
- all comments like this one 
https://github.com/Idle-Labs/idle-tranches/pull/28/files#r809718207 needs to be addressed before marking them as 'Resolved'

File: test/integration/harvest/IdleHarvestStrategy_DAI.js [hunk 12]
```diff
@@ -100,23 +101,32 @@ describe.only("Idle Harvest Strategy (DAI)", async () => {
 
   it("Deposit", async () => {
     console.log("Idle Harvest Strategy", IdleHarvestStrategy.address);
+    let balanceBefore = await rewardPoolContract.balanceOf(IdleHarvestStrategy.address);
     await DAI.connect(user).approve(IdleHarvestStrategy.address, AMOUNT_TO_TRANSFER);
     await IdleHarvestStrategy.connect(user).deposit(AMOUNT_TO_TRANSFER);
+    let balanceAfter = await rewardPoolContract.balanceOf(IdleHarvestStrategy.address);
```
Comment (bugduino) [inline test/integration/harvest/IdleHarvestStrategy_DAI.js:107]: Can you also add an assertion also for checking the increase of strategyTokens `user` (and the opposite in the redeem)

File: test/integration/harvest/IdleHarvestStrategy_USDT.js [hunk 13]
```diff
@@ -116,24 +117,33 @@ describe.only("Idle Harvest Strategy (USDT)", async () => {
...
     await USDT.connect(user).approve(IdleHarvestStrategy.address, AMOUNT_TO_TRANSFER);
     await IdleHarvestStrategy.connect(user).deposit(AMOUNT_TO_TRANSFER);
 
+    let balanceBefore = await rewardPoolContract.balanceOf(IdleHarvestStrategy.address);
     await IdleHarvestStrategy.connect(user).redeemUnderlying(toRedeem);
+    let balanceAfter = await rewardPoolContract.balanceOf(IdleHarvestStrategy.address);
+    expect(balanceBefore).gt(balanceAfter);
   });
 
   it("Price, apr and rewardTokens", async () => {
```
Comment (bugduino) [inline test/integration/harvest/IdleHarvestStrategy_USDT.js:149]: can you add some assertions here too? eg initially the apr should be 0 and after the first interaction with the strategy it should be > 0

File: .gitignore [hunk 14]
```diff
@@ -12,3 +12,4 @@ coverage
 bin/**
 soljson-v*
 audit*
+hardhat.config.js
```
Comment (bugduino) [inline .gitignore:15]: This should not be gitignored

File: contracts/strategies/harvest/IdleHarvestStrategy.sol [hunk 15]
```diff
@@ -162,15 +163,17 @@ contract IdleHarvestStrategy is Initializable, OwnableUpgradeable, ERC20Upgradea
     function redeemUnderlying(uint256 _amount) external returns (uint256) {
         if (_amount > 0) {
             uint256 _underlyingAmount = (_amount * oneToken) / price();
```
Comment (bugduino) [inline contracts/strategies/harvest/IdleHarvestStrategy.sol:165]: You can fetch `price()` only once

File: contracts/strategies/harvest/IdleHarvestStrategy.sol [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+    function redeem(uint256 _amount) external override onlyIdleCDO returns (uint256) {
+        if (_amount > 0) {
+            return _redeem(_amount, price());
+        }
+    }
+
+    /// @notice Redeem Tokens
+    /// @param _amount amount of underlying tokens to redeem
+    /// @return Amount of underlying tokens received
+    function redeemUnderlying(uint256 _amount) external returns (uint256) {
```
Comment (dantop114) [inline contracts/strategies/harvest/IdleHarvestStrategy.sol:163]: this method should be protected using `onlyCDO` right?

Comment (dantop114) [reply contracts/strategies/harvest/IdleHarvestStrategy.sol:163]: otherwise the receiver of funds in `_redeem` is `msg.sender`. this should not happen

**Review by dantop114 — CHANGES_REQUESTED** [review]
Strat looks good apart from comments by @bugduino on tests that were not solved I think.
Also there's a problem (I left a comment) with authorisation on the `redeemUnderlying` method.

**Review by massun-onibakuchi — CHANGES_REQUESTED** [review]

File: yarn.lock [hunk 17]
```diff
@@ -1770,6 +1770,11 @@
   resolved "https://registry.yarnpkg.com/@openzeppelin/contracts-upgradeable/-/contracts-upgradeable-4.0.0.tgz#11edb64933c43ab3eab2a84abe5e3ccf2981c4c7"
   integrity sha512-T5tO/KD++m+Ph74ppPPmNuhyrvNcsMDgQWt+pGshNJMsTf9UvmhBNyyOqVAL91UeuqDI0FHAbBV1+NnMg7ffFA==
 
+"@openzeppelin/contracts@3.4.1-solc-0.7-2":
+  version "3.4.1-solc-0.7-2"
+  resolved "https://registry.yarnpkg.com/@openzeppelin/contracts/-/contracts-3.4.1-solc-0.7-2.tgz#371c67ebffe50f551c3146a9eec5fe6ffe862e92"
+  integrity sha512-tAG9LWg8+M2CMu7hIsqHPaTyG4uDzjr6mhvH96LvOpLZZj6tgzTluBt+LsCf1/QaYrlis6pITvpIaIhE+iZB+Q==
```
Comment (massun-onibakuchi) [inline yarn.lock:1776]: why is this package added ? there is already "@openzeppelin/contracts@^4.4.2". maybe this part of the change should be revert.

**Review by massun-onibakuchi — CHANGES_REQUESTED** [review]
